firedrake/slate/slate.py

```diff
@@ -0,0 +1,619 @@
+
+    @property
+    def inv(self):
+        return Inverse(self)
+
+    @property
+    def T(self):
+        return Transpose(self)
+
+    def __add__(self, other):
+        if isinstance(other, TensorBase):
+            return Add(self, other)
+        else:
+            raise NotImplementedError("Operand type(s) for + not implemented: '%s' '%s'"
+                                      % (type(self), type(other)))
+
+    def __radd__(self, other):
+        # If other is not a TensorBase, raise NotImplementedError. Otherwise,
+        # delegate action to other.
+        if not isinstance(other, TensorBase):
+            raise NotImplementedError("Operand type(s) for + not implemented: '%s' '%s'"
+                                      % (type(other), type(self)))
+        else:
+            other.__add__(self)
+
+    def __sub__(self, other):
+        if isinstance(other, TensorBase):
+            return Sub(self, other)
+        else:
+            raise NotImplementedError("Operand type(s) for - not implemented: '%s' '%s'"
+                                      % (type(self), type(other)))
+
+    def __rsub__(self, other):
+        # If other is not a TensorBase, raise NotImplementedError. Otherwise,
+        # delegate action to other.
+        if not isinstance(other, TensorBase):
+            raise NotImplementedError("Operand type(s) for - not implemented: '%s' '%s'"
+                                      % (type(other), type(self)))
+        else:
+            other.__sub__(self)
+
+    def __mul__(self, other):
+        # if other is a ufl.Coefficient, return action
+        if isinstance(other, Coefficient):
+            return Action(self, other)
+        return Mul(self, other)
+
+    def __rmul__(self, other):
+        # If other is not a TensorBase, raise NotImplementedError. Otherwise,
+        # delegate action to other.
+        if not isinstance(other, TensorBase):
+            raise NotImplementedError("Operand type(s) for * not implemented: '%s' '%s'"
+                                      % (type(other), type(self)))
+        else:
+            other.__mul__(self)
+
+    def __neg__(self):
+        return Negative(self)
+
+    def __eq__(self, other):
+        """Determines whether two TensorBase objects are equal using their
+        associated keys.
+        """
+        return self._key() == other._key()
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def ufl_domain(self):
+        """This function returns a single domain of integration occuring
+        in the tensor.
+
+        The function will fail if multiple domains are found.
+        """
+        domains = self.ufl_domains()
+        assert all(domain == domains[0] for domain in domains), (
+            "All integrals must share the same domain of integration."
+        )
+        return domains[0]
+
+    def __str__(self):
+        """Returns a string representation."""
+        return self._output_string(self.prec)
```

Comment:
So every subclass must implement `_output_string` and `_key` (below).  Not in this pull request, but as a separate fix up, think about making `TensorBase` an [abstract base class](https://docs.python.org/2/library/abc.html) and document the required subclass methods/properties by annotating with `@abstractmethod` `@abstractproperty` as appropriate.